# Prelecture Homework
## 1. 

In [3]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

## 2. Chatbot session started
### 2. i. 

In [10]:
print(df.shape)
df

(391, 11)


row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

### 2. ii. General definitions of observation, variables, 

### Observations are individual data points/rows in a dataset. In this dataset, each row corresponds to one villager. 
### Variables are the attributes/columns in a dataset. Each variable in this dataset represents a charcteristic of the villagers (observations). They describe what is being measured for the observations. 

# 3.

In [16]:
df.info

<bound method DataFrame.info of      row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye

## 3. 

In [17]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [25]:
df['id'].value_counts()

id
admiral    1
mott       1
paula      1
patty      1
pate       1
          ..
eloise     1
elmer      1
ellie      1
elise      1
zucker     1
Name: count, Length: 390, dtype: int64

In [27]:
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [28]:
df.shape

(391, 11)

# 4.

In [31]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [32]:
df.shape

(891, 15)

In [34]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

#### In df.shape, it shows exactly the number of row (observation) entries there are, whereas in df.describe, it counted the number of valid entries there were. Since there were 177 missing entry values under 'age' it didn't count those, because they simply don't exist. Therefore it counted the values that were there (714).     891-177=714    Additionally, df.shape lists that there are 15 columns, but df.describe() only displays 6, and this is because df.describe() can only compute numerical values, thus making calculations for the 6 columns containing numerical values. 

# 5.

#### An attribute represents a characteristic of an object, thus you access it without brackets. They do not perform any computational actions (because they only hold static information). It tells you information that is already there.
#### However, a method performs an action. They take inputs and can perform calculations, retrieve data, manipulate data, etc. Methods calculate information that is there into something useful to you, depending on the operation. 

# Post-lecture homework
# 6.

In [35]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

#### 'Count' displays the number of non-missing values within each column. It shows how many actual data points there are in a given column
#### 'Mean' is the average of the values in a given column. It is calculated by adding all of the values and dividing the total by the count of values for the column displayed under 'count'. 
#### 'std', meaning standard deveation, is a measure of how spread out the values are around the mean. The higher the standard deviation, the more the values are spread out, and the lower it is, the more the values are clustered around the mean.
#### 'min', meaning minimum, is the smallest value in the column. 
#### '25%' This is the 25th percentile of the data, meaning 25% of the values in the column are below this value. It is also known as the lower quartile.
#### '50%', The 50th percentile, also known as the median, is the middle value when the data points are arranged in order. Half of the values are below the median and half are above.
#### '75%', The 75th percentile, meaning 75% of the values in the column are below this value.
#### 'max', meaning maximum is the biggest value in the column.

# 7.

#### An example of a "use case" in which using df.dropna() might be peferred over using del df['col'] is when you want to remove only rows of missing data, and not entire columns of missing data. This helps preserve as much of the data as possible. 
#### On the other hand, using del df['col'] is usually preferred when you want to discard of an entire column because it may consist almost entirely of missing values, or if the column is unnecesary to your analysis. 
#### By using del df['col'] first it can prevent the unnecessary deletion of rows that have missing values only in the irrelevant column. Then, it ensures that when you use df.dropna(), it will focus on the relevant data, without having to drop relevant data points.

In [4]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

#### Before: As shown above, before you drop the missing values, you can see that there are 177 missing values under 'age' and 688 missing values under 'deck'

In [44]:
df.dropna()
df=df.dropna()

In [53]:
df.isna().sum()

survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

In [48]:
print(df.shape)

(182, 15)


#### After: I used df.dropna, because i'm not sure if 'age' and/or 'deck' are important columns for an analysys. When using df.dropna, it dropped all the rows with missing values, thus leaving me with a dataset with 182 rows and 15 columns. I didn't have to drop any potentially important column information that may be useful for analysis, while still having enough rows of data for the information to be useful. 

# 8.

In [59]:
df.columns

Index(['survived', 'pclass', 'sex', 'sibsp', 'parch', 'fare', 'embarked',
       'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone'],
      dtype='object')

In [64]:
df.groupby("survived")["fare"].describe()

count       mean        std   min    25%      50%   75%       max
survived                                                                   
0          59.0  64.532131  62.108090  0.00  27.15  50.4958  79.2  263.0000
1         123.0  85.821107  81.843522  8.05  30.50  69.3000  93.5  512.3292

#### It grouped survivors vs non-survivors and counted several measures (by using the .describe() function) of the fare associated with each of the 2 possibilities (survivor/non survivor)

In [66]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

#### The function df.describe() has different values than df.groupby("col1")["col2"].describe() because df.groupby("col1")["col2"].describe() is specific to particular columns that the information is in, in relation to another column (grouped by values of specific columns). Whereas df.describe() is for the whole df, and is non-specific.

In [67]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [ ]:
#8.3 compare A-G whether chat is more efficient or google 

# 8.3

In [1]:
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

NameError: name 'pd' is not defined

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

####  A) in this issue chat knew right away how to fix the issue, much quicker than google

In [5]:
url = "titanics.csv"
df = pd.read_csv(url)
df.isna().sum()

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

#### B) ChatGPT noticed that a url was missing, but I still had to use google to search the web for an actual link to input as the URL, so the use of both together was the most useful

In [6]:
DF.shape

NameError: name 'DF' is not defined

#### C) ChatGPT knew what to do right away with the issue, whereas google did too but it couldn't give you instructions specific to your code.

In [7]:
df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

In [8]:
df.describe(

SyntaxError: incomplete input (3431639942.py, line 1)

#### D) ChatGPT realized the issue could be an open paranthesis right away in the first possible issue, whereas google just gave a link to a reddit page where you need to read for yourself

In [9]:
df.describle()

AttributeError: 'DataFrame' object has no attribute 'describle'

#### E) ChatGPT realized the issue right away, but google would again, take you to reddit where you would have to read through a lot of possibilities

In [14]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

In [16]:
df.groupby("Sex")["age"].describe()

KeyError: 'Sex'

#### F) ChatGPT realized the error right away, google could not understand it

In [19]:
df.groupby("sex")[age].describe()

NameError: name 'age' is not defined

##### G) ChatGPT noticed problem right away, google again could not understand the issue because it was not specific

In [ ]:
https://chatgpt.com/share/66e3b8bf-2404-8003-8eec-8cc9fa083cb6
https://chatgpt.com/share/66e3b8a3-d688-8003-a568-60c67f9b8fd9
    
# Summary:
#Using del df['col'] first can prevent the deletion of rows that have missing values only in the irrelevant column. It ensures that df.dropna() focuses on the relevant data, leading to cleaner and more efficient data handling.

#You said:
#Provide a summary of our chat session
#ChatGPT said:

#ChatGPT
#In this session, we covered several important topics related to pandas DataFrame manipulation and summary statistics:

#Identifying the Size of a DataFrame:
#You can use df.shape to find out the number of rows (observations) and columns (variables) in your dataset.
#Understanding the Terms "Observations" and "Variables":
#Observations refer to rows, which represent individual data points (like a villager in your dataset).
#Variables refer to columns, which represent attributes or characteristics of those observations.
#Summarizing DataFrame Columns:
#Methods like df.describe(), df.info(), df.nunique(), and df.value_counts() can help provide insights into your dataset, such as summary statistics, data types, and unique values in each column.
#Attributes vs. Methods in Pandas:
#We discussed the difference between attributes (e.g., df.shape), which return static properties of a DataFrame, and methods (e.g., df.describe()), which perform actions or calculations.
#Explanation of Summary Statistics:
#We reviewed the meanings of summary statistics like count, mean, std, min, 25%, 50%, 75%, and max, which describe the distribution and spread of numerical data in the DataFrame.
#Choosing Between df.dropna() and del df['col']:
#We explored a use case where using df.dropna() is better for preserving useful data while removing rows with missing values, compared to del df['col'], which completely removes a column.
#Why Order Matters When Using del and dropna() Together:
#Applying del df['col'] before df.dropna() prevents unnecessary row deletion and ensures that only relevant missing data is cleaned.
#Let me know if you need further clarification on any of these points!